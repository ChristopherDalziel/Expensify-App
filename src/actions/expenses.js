import { v4 as uuidv4 } from "uuid";
import database from "../firebase/firebase";

export const addExpense = (expense) => ({
  type: "ADD_EXPENSE",
  expense,
});

export const startAddExpense = (expenseData = {}) => {
  // returning a function, that works only because we are using thunk
  return (dispatch) => {
    // Setting up the default object
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0,
    } = expenseData;
    // Destructor our data
    const expense = { description, note, amount, createdAt };
    // Save our data
    database
      .ref("expenses")
      .push(expense)
      .then((ref) => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense,
          })
        );
      });
  };
};

export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id,
});

export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates,
});
