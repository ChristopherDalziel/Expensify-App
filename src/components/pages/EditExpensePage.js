import React from "react";

const EditExpensePage = (props) => {
  return (
    <>
      <h1>Edit Expense</h1>
      <p>Editing expense {props.match.params.id}</p>
    </>
  );
};

export default EditExpensePage;
