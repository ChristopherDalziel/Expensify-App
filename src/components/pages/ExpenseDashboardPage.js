import React from "react";

const ExpenseDashboardPage = () => {
  return <div>Dashboard Page</div>;
};

export default ExpenseDashboardPage;
