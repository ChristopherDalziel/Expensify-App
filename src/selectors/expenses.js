import moment from "moment";

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter((expense) => {
      // We are now using moment dates here to align with the component that is using date picker rather than manually inputted dates
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, "day")
        : true;

      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, "day")
        : true;

      const matchText = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());

      return startDateMatch && endDateMatch && matchText;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.createdAt < b.createdAt ? 1 : -1;
      }
      if (sortBy === "amount") {
        return a.amount < b.amount ? 1 : -1;
      }
    });
};

export default getVisibleExpenses;
