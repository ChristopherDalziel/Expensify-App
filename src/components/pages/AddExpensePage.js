import React from "react";
import { connect } from "react-redux";
import styled from "@emotion/styled";
import ExpenseForm from "../ExpenseForm";
import { startAddExpense } from "../../actions/expenses";

const AddExpensePageContainer = styled.div`
  margin: 10px;
`;

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    // The addExpense function we set up earlier takes an object with all of the values that the expense prop we're passing up is holding.
    this.props.startAddExpense(expense);
    this.props.history.push("/");
  };
  render() {
    return (
      <AddExpensePageContainer>
        <h1>Add Expense</h1>
        <ExpenseForm onSubmit={this.onSubmit} />
      </AddExpensePageContainer>
    );
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    startAddExpense: (expense) => dispatch(startAddExpense(expense)),
  };
};

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
