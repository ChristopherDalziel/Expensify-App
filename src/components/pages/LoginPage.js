import React from "react";
import { connect } from "react-redux";
import { startLogin } from "../../actions/auth";
import Button from "react-bootstrap/Button";
import styled from "@emotion/styled";

const LoginContainer = styled.main`
  height: 100vh;
  display: flex;
  justify-content: center;
  align-items: center;
`;

export const LoginPage = ({ startLogin }) => {
  return (
    <LoginContainer>
      <Button variant="outline-primary" onClick={startLogin}>
        Login
      </Button>
    </LoginContainer>
  );
};

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin()),
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
