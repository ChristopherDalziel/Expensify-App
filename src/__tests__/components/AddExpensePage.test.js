import React from "react";
import { shallow } from "enzyme";
import { AddExpensePage } from "../../components/pages/AddExpensePage";
import expenses from "../fixtures/expenses";

import toJSON from "enzyme-to-json";

let onSubmit, history, wrapper;

beforeEach(() => {
  onSubmit = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(<AddExpensePage onSubmit={onSubmit} history={history} />);
});

test("Should render add expense page correctly", () => {
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test("Should handle onSubmit", () => {
  wrapper.find("ExpenseForm").prop("onSubmit")(expenses[1]);
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(onSubmit).toHaveBeenLastCalledWith(expenses[1]);
});
