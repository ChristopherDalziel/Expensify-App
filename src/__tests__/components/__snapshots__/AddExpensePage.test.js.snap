// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render addExpense page correctly 1`] = `
<div>
  <h1>
    Add Expense
  </h1>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;

exports[`Should render startAddExpense page correctly 1`] = `
<div>
  <h1>
    Add Expense
  </h1>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
